File: dotnet/Nitro/Geometry/DoubleVector3.cs
namespace Nitro.Geometry
{
    public class DoubleVector3
    {
        public double X;
        
        public double Y;
        
        public double Z;

        public DoubleVector3()
        {
            X = Y = Z = 0;
        }

        public DoubleVector3(double x, double y, double z)
        {
            X = x;
            Y = y;
            Z = z;
        }

        public DoubleVector3(DoubleVector3 vec)
        {
            X = vec.X;
            Y = vec.Y;
            Z = vec.Z;
        }

        /// <summary>
        /// Returns the length of this vector
        /// </summary>
        public double Length()
        {
            return Math.Sqrt(X * X + Y * Y + Z * Z);
        }

        /// <summary>
        /// Returns |this| * |this|. Performs better than Length() and is sufficient
        /// for comparisions when e.g. sorting vectors by length.
        /// </summary>
        public double LengthSq()
        {
            return X * X + Y * Y + Z * Z;
        }

        /// <summary>
        /// Performs the dot product
        /// </summary>
        public double Dot(DoubleVector3 a)
        {
            return this.X * a.X + this.Y * a.Y + this.Z * a.Z;
        }

        /// <summary>
        /// Calculates distance from this point to the given one
        /// </summary>
        /// <returns>Distance</returns>
        public double DistanceTo(DoubleVector3 vector)
        {
            return Math.Sqrt((this.X - vector.X) * (this.X - vector.X) +
                (this.Y - vector.Y) * (this.Y - vector.Y) +
                (this.Z - vector.Z) * (this.Z - vector.Z));
        }

        /// <summary>
        /// Performs the cross product of this and c
        /// </summary>
        public DoubleVector3 Cross(DoubleVector3 c)
        {
            double x = this.Y * c.Z - this.Z * c.Y;
            double y = this.Z * c.X - this.X * c.Z;
            double z = this.X * c.Y - this.Y * c.X;

            return new DoubleVector3(x, y, z);
        }

        public bool IsColinear(DoubleVector3 a)
        {
            DoubleVector3 v1 = new DoubleVector3(this);
            DoubleVector3 v2 = new DoubleVector3(a);

            v1.Normalize();
            v2.Normalize();

            return (v1 - v2).LengthSq() < 0.001;
        }

        /// <summary>
        /// Calculates the Angle included between this and the given vector
        /// </summary>
        /// <returns>Angle in radians</returns>
        public double Angle(DoubleVector3 vector)
        {
            double numerator = this.Dot(vector);
            double denominator = this.Length() * vector.Length();

            return Math.Acos(numerator / denominator);
        }

        public override string ToString()
        {
            return String.Format("{0:0.00000}, {1:0.00000}, {2:0.00000}", X, Y, Z);
        }

        public override bool Equals(object? obj)
        {
            if (obj is not DoubleVector3)
                return false;

            var vector = (DoubleVector3)obj;
            return vector.X == X && vector.Y == Y && vector.Z == Z;
        }

        public override int GetHashCode()
        {
            return X.GetHashCode() + Y.GetHashCode() + Z.GetHashCode();
        }

        public static DoubleVector3 operator +(DoubleVector3 a, DoubleVector3 b)
        {
            return new DoubleVector3(a.X + b.X, a.Y + b.Y, a.Z + b.Z);
        }

        public static DoubleVector3 operator -(DoubleVector3 a)
        {
            return new DoubleVector3(-a.X, -a.Y, -a.Z);
        }

        public static DoubleVector3 operator -(DoubleVector3 a, DoubleVector3 b)
        {
            return new DoubleVector3(a.X - b.X, a.Y - b.Y, a.Z - b.Z);
        }

        public static DoubleVector3 operator *(DoubleVector3 a, double factor)
        {
            return new DoubleVector3(a.X * factor, a.Y * factor, a.Z * factor);
        }

        public static DoubleVector3 operator *(double factor, DoubleVector3 a)
        {
            return new DoubleVector3(a.X * factor, a.Y * factor, a.Z * factor);
        }

        public static double operator*(DoubleVector3 a, DoubleVector3 b)
        {
            return a.Dot(b);
        }

        public static DoubleVector3 operator /(DoubleVector3 a, double factor)
        {
            return new DoubleVector3(a.X / factor, a.Y / factor, a.Z / factor);
        }

        public static bool operator ==(DoubleVector3? a, DoubleVector3? b)
        {
            return a?.X == b?.X && a?.Y == b?.Y && a?.Z == b?.Z;
        }

        public static bool operator !=(DoubleVector3? a, DoubleVector3? b)
        {
            return !(a == b);
        }

        public static implicit operator DoubleVector2(DoubleVector3 vector)
        {
            return new DoubleVector2(vector.X, vector.Y);
        }

        public static implicit operator DoubleVector3(DoubleVector2 vector)
        {
            return new DoubleVector3(vector.X, vector.Y, 0);
        }

        /// <summary>
        /// Performs in-place matrix multiplication
        /// </summary>
        /// <param name="transform">Matrix to multiply with</param>
        public DoubleVector3 Transform(DoubleMatrix transform)
        {
            double tempX = X * transform.M11 + Y * transform.M21 + Z * transform.M31 + transform.M41;
            double tempY = X * transform.M12 + Y * transform.M22 + Z * transform.M32 + transform.M42;
            double tempZ = X * transform.M13 + Y * transform.M23 + Z * transform.M33 + transform.M43;
            double tempW = X * transform.M14 + Y * transform.M24 + Z * transform.M34 + transform.M44;

            X = tempX / tempW;
            Y = tempY / tempW;
            Z = tempZ / tempW;

            return this;
        }

        /// <summary>
        /// Performs in-place addition
        /// </summary>
        /// <param name="a"></param>
        public DoubleVector3 Add(DoubleVector3 a)
        {
            this.X += a.X;
            this.Y += a.Y;
            this.Z += a.Z;

            return this;
        }

        /// <summary>
        /// Performs in-place scalar multiplication
        /// </summary>
        /// <param name="scalar"></param>
        public DoubleVector3 Multiply(double scalar)
        {
            this.X *= scalar;
            this.Y *= scalar;
            this.Z *= scalar;

            return this;
        }

        /// <summary>
        /// Normalizes this vector. Vector remains unchanged in case length is 0.
        /// </summary>
        public void Normalize()
        {
            double length = Length();

            if (length == 0)
                return;

            X /= length;
            Y /= length;
            Z /= length;
        }
    }
}

File: web/webmap/src/geometry/DoubleVector3.ts
import { DoubleMatrix } from "./DoubleMatrix";

export class DoubleVector3 {
    constructor(public x: number = 0, public y: number = 0, public z: number = 0) {
    }

    public clone(): DoubleVector3 {
        return new DoubleVector3(this.x, this.y, this.z);
    }

    public length(): number {
        return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
    }

    public lengthSq(): number {
        return this.x * this.x + this.y * this.y + this.z * this.z;
    }

    public dot(a: DoubleVector3): number {
        return this.x * a.x + this.y * a.y + this.z * a.z;
    }

    public distanceTo(vector: DoubleVector3): number {
        return Math.sqrt(
            (this.x - vector.x) * (this.x - vector.x) +
            (this.y - vector.y) * (this.y - vector.y) +
            (this.z - vector.z) * (this.z - vector.z)
        );
    }

    public distanceToSq(vector: DoubleVector3): number {
        return (this.x - vector.x) * (this.x - vector.x) +
            (this.y - vector.y) * (this.y - vector.y) +
            (this.z - vector.z) * (this.z - vector.z);
    }

    public cross(c: DoubleVector3): DoubleVector3 {
        const x = this.y * c.z - this.z * c.y;
        const y = this.z * c.x - this.x * c.z;
        const z = this.x * c.y - this.y * c.x;

        return new DoubleVector3(x, y, z);
    }

    public isColinear(a: DoubleVector3): boolean {
        const v1 = new DoubleVector3(this.x, this.y, this.z);
        const v2 = new DoubleVector3(a.x, a.y, a.z);

        v1.normalize();
        v2.normalize();

        return (v1.subtract(v2).lengthSq() < 0.001);
    }

    public angle(vector: DoubleVector3): number {
        const numerator = this.dot(vector);
        const denominator = this.length() * vector.length();

        return Math.acos(numerator / denominator);
    }

    public toString(): string {
        return `${this.x.toFixed(5)}, ${this.y.toFixed(5)}, ${this.z.toFixed(5)}`;
    }

    public equals(obj: any): boolean {
        if (!(obj instanceof DoubleVector3)) {
            return false;
        }

        const vector = obj as DoubleVector3;
        return vector.x === this.x && vector.y === this.y && vector.z === this.z;
    }

    public add(a: DoubleVector3): DoubleVector3 {
        this.x += a.x;
        this.y += a.y;
        this.z += a.z;

        return this;
    }

    public multiply(scalar: number): DoubleVector3 {
        this.x *= scalar;
        this.y *= scalar;
        this.z *= scalar;

        return this;
    }

    public normalize(): void {
        const length = this.length();

        if (length === 0) {
            return;
        }

        this.x /= length;
        this.y /= length;
        this.z /= length;
    }

    public subtract(a: DoubleVector3): DoubleVector3 {
        return new DoubleVector3(this.x - a.x, this.y - a.y, this.z - a.z);
    }

    public static add(a: DoubleVector3, b: DoubleVector3): DoubleVector3 {
        return new DoubleVector3(a.x + b.x, a.y + b.y, a.z + b.z);
    }

    public static negate(a: DoubleVector3): DoubleVector3 {
        return new DoubleVector3(-a.x, -a.y, -a.z);
    }

    public static subtract(a: DoubleVector3, b: DoubleVector3): DoubleVector3 {
        return new DoubleVector3(a.x - b.x, a.y - b.y, a.z - b.z);
    }

    public static multiply(a: DoubleVector3, factor: number): DoubleVector3 {
        return new DoubleVector3(a.x * factor, a.y * factor, a.z * factor);
    }

    public transform(transform: DoubleMatrix): DoubleVector3 {
        const tempX =
            this.x * transform.M11 +
            this.y * transform.M21 +
            this.z * transform.M31 +
            transform.M41;
        const tempY =
            this.x * transform.M12 +
            this.y * transform.M22 +
            this.z * transform.M32 +
            transform.M42;
        const tempZ =
            this.x * transform.M13 +
            this.y * transform.M23 +
            this.z * transform.M33 +
            transform.M43;
        const tempW =
            this.x * transform.M14 +
            this.y * transform.M24 +
            this.z * transform.M34 +
            transform.M44;

        this.x = tempX / tempW;
        this.y = tempY / tempW;
        this.z = tempZ / tempW;

        return this;
    }
}

export const origin = new DoubleVector3(0, 0, 0);
